File: Correct/2학년/F. 소수의 신 희성/main.py
from itertools import permutations

def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5)+1):
        if n % i == 0:
            return False
    return True

N, M = map(int, input().split())
V = list(map(int, input().split()))

seen = set()
count = 0

for k in range(1, M+1):
    for p in permutations(V, k):
        num = int(''.join(map(str, p)))
        if num not in seen:
            seen.add(num)
            if is_prime(num):
                count += 1

print(count)

File: Correct/2학년/F. 소수의 신 희성/main.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <cmath>
using namespace std;

bool is_prime(int n) {
    if (n < 2) return false;
    for (int i = 2; i*i <= n; ++i)
        if (n % i == 0)
            return false;
    return true;
}

void backtrack(vector<int>& V, vector<bool>& used, vector<int>& path, int M, set<int>& seen, int& count) {
    if (!path.empty() && path.size() <= M) {
        int num = 0;
        for (int d : path) num = num * 10 + d;
        if (seen.find(num) == seen.end()) {
            seen.insert(num);
            if (is_prime(num)) ++count;
        }
    }
    if (path.size() == M) return;

    for (int i = 0; i < V.size(); ++i) {
        if (!used[i]) {
            used[i] = true;
            path.push_back(V[i]);
            backtrack(V, used, path, M, seen, count);
            path.pop_back();
            used[i] = false;
        }
    }
}

int main() {
    int N, M;
    cin >> N >> M;
    vector<int> V(N);
    for (int i = 0; i < N; ++i) cin >> V[i];

    vector<bool> used(N, false);
    vector<int> path;
    set<int> seen;
    int count = 0;

    backtrack(V, used, path, M, seen, count);
    cout << count << '\n';
    return 0;
}